### ISSUE TITLE
[C#] Implement new Concept Exercise: chars

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#960]
This issue describes how to implement the `chars` concept exercise for the C# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student the basics of the Concept of Chars in [C#][chars-docs].

## Learning objectives

- Know of the existence of the `char` type.
- Know what a char represents (a Unicode character).
- Know how to define a char.
- Know how to access a char in a string by index.
- Know of some basic char methods (like converting to uppercase).
- Know that chars are immutable.
- Know how to use a `for` loop.

## Out of scope

- Converting an integer to a character and vice versa.

## Concepts

This Concepts Exercise's Concepts are:

- `chars`: know of the existence of the `char` type; know that a char represents; know how to define a char; know how to access a char in a string by index; know of some basic char methods (like converting to uppercase).
- `for-loops`: know how to use a `for` loop.

## Prequisites

This Concept Exercise's prerequisites Concepts are:

- `strings`: know of the `string` type that will be iterated over and accessed by index.

### Hints

- [Chars documentation][chars-docs]: reference documentation for `char`.
- [Chars tutorial][chars-tutorial]: basic tutorial on how to work with chars.

### After

- [Chars documentation][chars-docs]: reference documentation for `char`.
- [Chars tutorial][chars-tutorial]: basic tutorial on how to work with chars.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

To implement this exercise, please [follow these instructions](https://github.com/exercism/v3/blob/master/languages/csharp/reference/implementing-a-concept-exercise.md).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/csharp/reference/README.md#reference-docs
[reference-char]: https://github.com/exercism/v3/blob/master/reference/types/char.md
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer
[exercise-example]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/numbers-floating-point
[design-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/design.md
[config.json-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/config.json
[concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[chars-docs]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/char
[chars-tutorial]: https://csharp.net-tutorials.com/data-types/the-char-type/

**Comment by mikedamay** [comment]
Can you say more about `for` loops.  Am I blanking on some special relationship between `for` and `char`?  Why are for loops not a pre-requisite?

**Comment by mikedamay** [comment]
Is boxing out of scope?  I would think that after.md would be a good place for it.

**Comment by ErikSchierboom** [comment]
> Can you say more about for loops. Am I blanking on some special relationship between for and char? Why are for loops not a pre-requisite?

I struggled a bit to come up with an exercise that was about chars but that was also in any way interesting. I then came up with the idea that we could have students iterate over the chars in a string using a `for` loop. Does that make sense or not? We could omit the `for` loop, but then the exercise might be very dull, so maybe then we should replace it with a different concept. 

Thinking about it a bit more, it might make more sense to introduce the `for` loop in the `arrays` exercise, which already introduces the `foreach` loop. What do you think about that?

> Is boxing out of scope? I would think that after.md would be a good place for it.

My gut feeling is that boxing would be probably best be discussed in another exercise, for example in the exercise that teaches about casting between different class types. What was your reasoning for wanting to include it here?

**Comment by mikedamay** [comment]
for-loops:

I had not considered the possibility of an exercise introducing disparate (albeit connected) concepts like chars and for loops.  One disadvantage of this approach is when using the exercises as a reference resource (either as a student of a maintainer) it will make it much less obvious where to look for relevant material.

I had not thought beyond the idea that there would be a `for` exercise or, given that this is rather fine grained, perhaps a loops exercise.

I certainly have no objection to the char exercise including `for` as a pre-requisite.  I think that your suggestion of `array` is a better place for `for` if we are going to use combinations in this way.

boxing:

I have frankly lost track of the significance of boxing.  I know that whenever I revert to Java I can see what a headache it can be.  I think that the experience has scarred me.  It doesn’t seem that, in a world of reified generics, boxing is terribly important.

I was not thinking of introducing boxing as a “concept" rather when discussing a “simple” type, in the interests of a rounded account, the aliased type might be mentioned, in our case System.Char, together with a nod towards the role of “boxing”.   If I pick this up I would propose putting something in after.md along these lines and perhaps it’s not important enough even to mention as being out of scope.

**Comment by ErikSchierboom** [comment]
> I had not considered the possibility of an exercise introducing disparate (albeit connected) concepts like chars and for loops. One disadvantage of this approach is when using the exercises as a reference resource (either as a student of a maintainer) it will make it much less obvious where to look for relevant material.

In general we try to avoid introducing multiple (disparate) concepts in one exercise, but in some cases it does make sense:

- When the concepts are closely related. For example foreach loops and arrays (being the first collection type) go well with each other. Extension methods and static classes could also be a good combination, and there are probably more. I do expect that for more advanced concepts, this will rarely be the case.
- It can make relatively basic exercise a bit more interesting. For some basic concepts (which includes char I think) are incredibly hard to design a slightly interesting exercise for. In these cases, introducing a (related) concept can greatly help here.

> I was not thinking of introducing boxing as a “concept" rather when discussing a “simple” type, in the interests of a rounded account, the aliased type might be mentioned, in our case System.Char, together with a nod towards the role of “boxing”. If I pick this up I would propose putting something in after.md along these lines and perhaps it’s not important enough even to mention as being out of scope.

Ah that sounds reasonable! I also don't think boxing is as much of an issue in C# as it is in Java (I frequently encountered its effects there), but let's see if it makes sense when referring to `System.Char` (which is an excellent point).

**Comment by ErikSchierboom** [comment]
> I think that your suggestion of array is a better place for for if we are going to use combinations in this way.

Shall I create an issue for it?

**Comment by ErikSchierboom** [comment]
@mikedamay I've created [a PR to introduce for-loops in the arrays exercise](https://github.com/exercism/v3/pull/1539). Let me know what you think.

**Comment by mikedamay** [comment]
I will give "char" a go.  To pick up on your idea - making identiers comply with naming conventions would seem to work.

**Comment by ErikSchierboom** [comment]
Great! Will be good to have a `chars` exercise. Could you start with a basic draft PR that has just the following three files:

- Example.cs
- `.docs/instructions.md`
- `.meta/design.md` or `.docs/introduction.md` (I prefer the former, but feel free to choose which of the two you prefer)

The PR would then look something like this: https://github.com/exercism/v3/pull/1486/files (albeit in C#).

**Comment by mikedamay** [comment]
@ErikSchierboom 

> Could you start with a basic draft PR that has just the following three files:

Please let me know the reason for this request.  If it is a question of spliting tasks into more manageable pieces and allowing for a measure of specialisation as discussed in last week's video call then I would suggest that some flexibility is in order.  This approach does not fit my workflow (happy to provide detais).

If it's a question of getting some communal participation earlier in the process then I would suggest a fuller discussion on the concept creation issue.

Let me know what you think.

**Comment by ErikSchierboom** [comment]
The reason for this request is just to check if the PR is going in the right direction. We've had PR's where a fully-fleshed exercise was PR'ed that turned out to require quite some modification. It would have been easier if the PR started with just the basics, and getting that right first. I don't necessarily mind the bigger PR's, but there will likely be more churn and it makes reviewing a bit harder.

> [...] I would suggest a fuller discussion on the concept creation issue.

I'm fine with that too. Did you have an idea on what you wanted the exercise to do? Which things will students be implementing?

**Comment by ErikSchierboom** [comment]
@mikedamay I don't know if it is any use for you, but Elixir is creating a [charlist exercise](https://github.com/exercism/v3/pull/1560#pullrequestreview-422407165). It is definitely not the same as working with regular chars, but you might find inspiration in the story.

**Comment by mikedamay** [comment]
@ErikSchierboom If you get a chance could you make a comment on [PR 1556[((https://github.com/exercism/v3/pull/1556).  I can then continue with that.

**Comment by ErikSchierboom** [comment]
@mikedamay Sure, I hadn't fully worked through my list of notifications :)

**Comment by mikedamay** [comment]
@ErikSchierboom Classic concurrency issue!

The elixir exercise would have done nicely.  It's more realistic than mine but programmers and identifiers go together reasonably well so I'm inclined to leave it.

**Comment by ErikSchierboom** [comment]
Haha, yes! I think the identifiers theme does make sense and will resonate with students.

**Comment by mikedamay** [comment]
@ErikSchierboom 

> Could you start with a basic draft PR that has just the following three files:
Example.cs
    .docs/instructions.md
    .meta/design.md or .docs/introduction.md (I prefer the former, but feel free to choose which of the two you prefer)

If you are formalsing this, maybe the following sort of idea/wording might be considered.  I appreciate that in many cases the wizard will supersede this approach so it may not be worth it.

"To start with, please submit a draft containing at least the following files:
- Example implmentation (Example.cs, example.js etc.)
- docs/instructions.md
- .meta/design.md or .docs/introduction.md (the former is preferable, but feel free to choose which of the two you prefer)

Intially only these files will be reviewed (in broad terms) so it might be adviseable to delay completion of the other files until the draft has been reviewed."

**Comment by ErikSchierboom** [comment]
@mikedamay That sounds like a nice way of wording it. 

P.S. I'll be looking at this PR shortly, I had a day off yesterday (speaking at NDC Oslo on the C# analyzer/representer/test runner).

**Comment by mikedamay** [comment]
Closed by #1556
